src/KernelStrategy.sol

```diff
@@ -92,11 +92,52 @@ contract KernelStrategy is Vault {
      * @return maxAssets The maximum amount of assets.
      */
     function maxWithdraw(address owner) public view override returns (uint256 maxAssets) {
-        if (paused()) {
+        maxAssets = _maxWithdrawAsset(asset(), owner);
+    }
+
+    /**
+     * @notice Returns the maximum amount of assets that can be withdrawn for a specific asset by a given owner.
+     * @param asset_ The address of the asset.
+     * @param owner The address of the owner.
+     * @return maxAssets The maximum amount of assets.
+     */
+    function maxWithdrawAsset(address asset_, address owner) public view returns (uint256 maxAssets) {
+        maxAssets = _maxWithdrawAsset(asset_, owner);
+    }
+
+    /**
+     * @notice Internal function to get the maximum amount of assets that can be withdrawn by a given owner.
+     * @param asset_ The address of the asset.
+     * @param owner The address of the owner.
+     * @return maxAssets The maximum amount of assets.
+     */
+    function _maxWithdrawAsset(address asset_, address owner) internal view virtual returns (uint256 maxAssets) {
+        if (paused() || !_getAssetStorage().assets[asset_].active) {
             return 0;
         }
 
-        (maxAssets,) = _convertToAssets(asset(), balanceOf(owner), Math.Rounding.Floor);
+        uint256 availableAssets = _availableAssets(asset_);
+
+        maxAssets = previewRedeemAsset(asset_, balanceOf(owner));
+
+        maxAssets = availableAssets < maxAssets ? availableAssets : maxAssets;
```

Comment:
@danoctavian I've modeled this code based on [maxWithdraw in BaseVault.sol](https://github.com/yieldnest/yieldnest-vault/blob/74d53eccc291e1fcced590b40dcff62e2ad98653/src/BaseVault.sol#L159)

Here, availableAssets is returned without reducing fees.. This might be a problem in some cases.

Please take a look at https://github.com/yieldnest/yieldnest-kernel-lrt/pull/64 where I've tried to fix this.

IMO we should fix this in BaseVault.sol too.

Reply:
i don't think it's a problem

That's the maximum amount of assets you can redeem

The way we take your fee in the withdraw() flow is by burning more of your shares, not by taking some of your assets since ERC4626 needs to give you whatever you asked for by convention

if i say i want to withdraw 100 units, you'll give me 100 units to comply with the standard but you'll burn more of my shares to charge me a fee

Reply:
Okay you are right. Thanks for the clarification, then can we close #64 ?